tests/test_lockable_component.cairo

```diff
@@ -0,0 +1,152 @@
+        .deploy(@erc721_constructor_calldata)
+        .unwrap();
+
+    // deploy recipient contract
+    let account_contract = declare("SimpleAccount").unwrap();
+    let (recipient, _) = account_contract
+        .deploy(
+            @array![883045738439352841478194533192765345509759306772397516907181243450667673002]
+        )
+        .unwrap();
+
+    // mint a new token
+    let dispatcher = IERC721Dispatcher { contract_address: erc721_contract_address };
+    dispatcher.mint(recipient, 1.try_into().unwrap());
+
+    // deploy account contract
+    let account_contract = declare("AccountPreset").unwrap();
+    let mut acct_constructor_calldata = array![erc721_contract_address.try_into().unwrap(), 1, 0];
+    let (account_contract_address, _) = account_contract
+        .deploy(@acct_constructor_calldata)
+        .unwrap();
+
+    (account_contract_address, erc721_contract_address)
+}
+
+
+#[test]
+fn test_lockable_owner() {
+    let (contract_address, erc721_contract_address) = __setup__();
+    let acct_dispatcher = IAccountDispatcher { contract_address: contract_address };
+
+    let token_dispatcher = IERC721Dispatcher { contract_address: erc721_contract_address };
+    let owner = acct_dispatcher.owner();
+    let token_owner = token_dispatcher.ownerOf(1.try_into().unwrap());
+
+    start_cheat_caller_address(contract_address, token_owner);
+
+    assert(owner == token_owner, 'invalid owner');
+    stop_cheat_caller_address(contract_address);
+}
+#[test]
+fn test_lockable() {
+    let (contract_address, _) = __setup__();
+    let acct_dispatcher = IAccountDispatcher { contract_address: contract_address };
+
+    let owner = acct_dispatcher.owner();
+
+    start_cheat_caller_address(contract_address, owner);
+
+    let lockable_dispatcher = ILockableDispatcher { contract_address };
+
+    lockable_dispatcher.lock(40);
+    let check_lock = lockable_dispatcher.is_lock();
+
+    assert(check_lock == true, 'Account Not Lock');
+    stop_cheat_caller_address(contract_address);
+}
+#[test]
+fn test_lockable_emits_event() {
+    let (contract_address, _) = __setup__();
+
+    let acct_dispatcher = IAccountDispatcher { contract_address: contract_address };
+
+    let owner = acct_dispatcher.owner();
+
+    // spy on emitted events
+    let mut spy = spy_events();
+
+    start_cheat_caller_address(contract_address, owner);
+
+    // call the lock function
+    let lockable_dispatcher = ILockableDispatcher { contract_address: contract_address };
+
+    lockable_dispatcher.lock(40);
+
+    // check events are emitted
+    spy
+        .assert_emitted(
+            @array![
+                (
+                    contract_address,
+                    LockableComponent::Event::AccountLocked(
+                        LockableComponent::AccountLocked {
+                            account: owner, locked_at: get_block_timestamp(), lock_until: 40
+                        }
+                    )
+                )
+            ]
+        );
+}
+
```

Comment:
include tests to:

1. check that you can't execute when locked. [example here.](https://github.com/horuslabsio/TBA/blob/main/tests/test_account.cairo#L268)
2. check that you cannot upgrade when locked. [example here](https://github.com/horuslabsio/TBA/blob/main/tests/test_account.cairo#L309).
3. check that you can not lock if you are not the owner. check that it throws the right error. [example here](https://github.com/horuslabsio/TBA/blob/main/tests/test_account.cairo#L334).
4. check that you can not lock if already locked. it throws the right error.
5. check that it unlocks once the lock duration ends.

Hint: to check for execute and upgrade, you need to update their implementations to include a check for account is not locked.

Reply:
implementing LockableComponent inside Account component will create some sort of circular dependencies

don't you think so 

`impl Lockable: LockableComponent::HasComponent<TContractState>`

Reply:
not inside the Account component, but inside the preset. the preset is an example base account using all the other components. typically if you decide to use the Lockable component, you want to protect both your upgrade and execute function with the lockable. dyg?

Reply:
i see

got it